asm-utils/src/data_parser.rs

```diff
@@ -14,76 +16,134 @@ pub enum DataValue {
-                    ".size",
-                    [Argument::Expression(Expression::Symbol(name)), Argument::Expression(Expression::Number(n))],
-                ) if Some(name.as_str()) == current_label => {
-                    let label = current_label.unwrap().to_string();
-                    objects
-                        .entry(current_label.unwrap().into())
-                        .and_modify(|entry| {
-                            let size: usize = entry.iter().map(|v| v.size()).sum();
-                            assert!(
-                                size as i64 == *n,
-                                "Invalid size for data object {name}: computed: {size} vs. specified: {n}"
-                            );
-                        })
-                        .or_insert_with(|| {
-                            object_order.push(label);
-                            assert!(*n == 0, "Nonzero size for object without elements: {name}");
-                            Default::default()
-                        });
+                    ".zero" | ".ascii" | ".asciz" | ".dword" | ".word" | ".half" | ".hword"
+                    | ".short" | ".byte",
+                    args,
+                ) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .extend(extract_data_value(dir.as_str(), args));
+                    } else {
+                        // This is most likely debug data.
+                    }
+                }
+                (".balign", [Argument::Expression(Expression::Number(byte_size))]) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .push(DataValue::Alignment(*byte_size as usize));
+                    }
+                }
+                (".p2align", [Argument::Expression(Expression::Number(pow_of_2))]) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .push(DataValue::Alignment((1 << pow_of_2) as usize));
+                    }
+                }
+                (".balign" | ".p2align", _) => {
+                    // TODO: implement the optional arguments of .balign and .p2align
+                    unimplemented!()
                 }
                 _ => {}
             },
             _ => {}
         }
     }
-    (objects, object_order)
+    data.sections
+}
+
+fn is_data_section<R: Register, F: FunctionOpKind>(arg: &Argument<R, F>) -> bool {
+    let full_name = match arg {
+        Argument::StringLiteral(name) => name.as_slice(),
+        Argument::Expression(Expression::Symbol(name)) => name.as_bytes(),
+        _ => return false,
+    };
+
+    // split the initial '.' out
+    let name = &full_name[(1 + full_name.iter().position(|c| *c == b'.').unwrap())..];
+
+    // isolate name until next '.'
+    let name = &name[..(name.iter().position(|e| *e == b'.').unwrap_or(name.len()))];
+
```

Comment:
gpt says

```rust
let parts: Vec<&str> = full_name.split('.').collect();
let name = parts.get(1).unwrap_or(&"");
```

Reply:
That is not an `&str`. I used `&[u8]` because that is the common denominator between `String` and `Vec<u8>` of the two different enum alternatives that could reach this code.

Besides, this code sucks. Searches the entire string and allocates a new `Vec`, and is not even shorter.

Reply:
I can do this:

```rs
let next_dot = name.iter().position(|e| *e == b'.');
let name = &name[..next_dot.unwrap_or(name.len())];
```

Reply:
sure, just kinda still surprised there's nothing better

Reply:
fixed